Generalize result values of all Indicators by introducing a result class value

matthiasschaub opened the pull request:
### Description
Every Indicator determines a result class value for itself.
The result class value range is 1 up to 5 and maps to the result labels.

### Corresponding issue
Closes #321 

### New or changed dependencies
- 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] How to choose result classes which are resolving to the same label? (e.g. 4 or 5 for green)
    - https://github.com/GIScience/ohsome-quality-analyst/issues/371

joker234 commented on workers/ohsome_quality_analyst/base/indicator.py:
```diff
@@ -41,19 +38,28 @@ class Result:
-        label (str): Traffic lights like quality label
-        value (float): The result value as float ([0, 1])
-        description (str): Description of the result
+        label (str): Traffic lights like quality label: Green, yellow or red. The value
+            is determined by the result classes
```
```suggestion
        label (str): Traffic lights like quality label: `green`, `yellow` or `red`. The
            value is determined by the result classes
```

joker234 commented on workers/ohsome_quality_analyst/base/indicator.py:
```diff
@@ -41,19 +38,28 @@ class Result:
+        label (str): Traffic lights like quality label: Green, yellow or red. The value
+            is determined by the result classes
+        value (float): The result value
+        class_ (int): The result class. An integer between 1 and 5. It maps to the
```
I'm not really a fan of using the `…_` notation for parameters/variables that are used outside the class. Is there a better option? `result_class` maybe? or `label_class`?

joker234 commented on workers/ohsome_quality_analyst/base/indicator.py:
```diff
@@ -41,19 +38,28 @@ class Result:
+    timestamp_osm: Optional[datetime] = None
+    value: Optional[float] = None
+    class_: Optional[Literal[1, 2, 3, 4, 5]] = None
+    description: str = ""
+    svg: str = ""
```
Since `description` and `svg` are both set in the `__init__` function, I would not set them with default values here.

matthiasschaub replied:
Changed to no default

joker234 commented on workers/ohsome_quality_analyst/base/indicator.py:
```diff
@@ -41,19 +38,28 @@ class Result:
+
+    @property
+    def label(self) -> Literal["green", "yellow", "red", "undefined"]:
+        labels = {1: "red", 2: "yellow", 3: "yellow", 4: "green", 5: "green"}
```
TODO: add this definition (numbers to color) to the docs

matthiasschaub replied:
added to dev docs

joker234 commented on workers/ohsome_quality_analyst/geodatabase/client.py:
```diff
@@ -127,7 +127,7 @@ async def load_indicator_results(
     indicator.result.timestamp_oqt = query_result["timestamp_oqt"]
     indicator.result.timestamp_osm = query_result["timestamp_osm"]
-    indicator.result.label = query_result["result_label"]
+    indicator.result.class_ = query_result["result_class_"]
```
If we wouldn't change `class_` (see earlier comment), do you really want to use `class_` in the `query_result` dict?

matthiasschaub replied:
I kept the internal python attribute name `class_` and changed it to `class` for all names outside of the python code (e.g. database or geojson field name)

joker234 reviewed (commented):
In general pretty good. Please create an issue for each indicator to adjust to 5 classes instead of 3.

matthiasschaub commented:
> In general pretty good. Please create an issue for each indicator to adjust to 5 classes instead of 3.

I created an issue for this.
https://github.com/GIScience/ohsome-quality-analyst/issues/371

matthiasschaub commented on workers/tests/integrationtests/test_base_indicator.py:
```diff
@@ -51,3 +50,11 @@ def test_data_property(self):
+
+class TestBaseResult(unittest.TestCase):
+    def test_label(self):
+        result = Result()
```
fixed

matthiasschaub commented:
@joker234 thank you for the review. I updated the branch and resolved existing conlicts (`changelog`, `test_base_indicator.py` and `test_base_report.py`. Also I rewrote two small tests to use pytest and not unittest and removing the depedency on the database of those two tests (`test_base_indicator.py` and `test_base_report.py`).
Coul you please have another look?
